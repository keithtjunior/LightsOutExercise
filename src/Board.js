import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 * 
 **/

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard());

  /** game board: nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    for (let y=0; y<nrows; y++) {
      let row = []
      for (let x=0; x<ncols; x++) {
        row.push(Math.random() >= chanceLightStartsOn)
      }
      initialBoard.push(row)
    }
    return initialBoard;
  }

  // check board for all false values
  function hasWon() {
    return board.every(i => Array.isArray(i) ? i.every(j => !j) : !i)
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if coord on board flip it
        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // make a (deep) copy of the oldBoard
      let deepCopy = JSON.parse(JSON.stringify(oldBoard));

      // flip value of cell and the cells around it
      flipCell(y, x, deepCopy);
      flipCell(y, x+1, deepCopy); // right
      flipCell(y-1, x, deepCopy); // below
      flipCell(y, x-1, deepCopy); // left
      flipCell(y+1, x, deepCopy); // top
      
      return deepCopy;
    });
  }

  return (
      <div>
        { hasWon() ? 
          <h1>You Won!</h1> : 
          <table>
            <tbody>
              {board.map((row,y) => (
                <tr key={`row-${y}`}>
                  {row.map((col, x) => (
                    <Cell
                      key={`${y}-${x}`}
                      coord={`${y}-${x}`}
                      isLit={col}
                      flipCellsAroundMe={() => flipCellsAround(`${y}-${x}`)} />
                    )
                  )}
                </tr>
                )
              )}
            </tbody>
          </table>
        }
			</div>
  );
}
Board.defaultProps = {
  nrows: 5,
  ncols: 5, 
  chanceLightStartsOn: 0.4
}

export default Board;
