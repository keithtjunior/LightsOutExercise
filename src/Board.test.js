import { render, queryByAttribute, fireEvent } from "@testing-library/react";
import Board from "./Board";

// mocks Math.random value
beforeEach(function() {
    jest
    .spyOn(Math, "random")
    .mockReturnValue(0.9);
});
  
afterEach(function() {
    Math.random.mockRestore();
});

it("renders without crashing", function() {
    render(<Board />);
  });
  
it("matches snapshot: all true", function() {
    const { asFragment } = render(<Board />);
    expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot: all false: ie. 'You Won!'", function() {
    const { asFragment } = render(<Board chanceLightStartsOn={0.99} />);
    expect(asFragment()).toMatchSnapshot();
});

it("flips value of child (Cell) componet and its surrounding cells", function() {
    const getById = queryByAttribute.bind(null, 'id');
    const { container } = render(<Board />);
    const litCol1 = getById(container, '0-0');
    const litCol2 = getById(container, '0-1');
    const litCol3 = getById(container, '1-0');
    

    expect(litCol1).toHaveClass('Cell-lit'); 
    expect(litCol2).toHaveClass('Cell-lit'); 
    expect(litCol3).toHaveClass('Cell-lit'); 

    fireEvent.click(litCol1);

    expect(litCol1).not.toHaveClass('Cell-lit'); 
    expect(litCol2).not.toHaveClass('Cell-lit'); 
    expect(litCol3).not.toHaveClass('Cell-lit'); 
});