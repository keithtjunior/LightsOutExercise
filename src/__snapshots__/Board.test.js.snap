// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot: all false: ie. 'You Won!' 1`] = `
<DocumentFragment>
  <div>
    <h1>
      You Won!
    </h1>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot: all true 1`] = `
<DocumentFragment>
  <div>
    <table>
      <tbody>
        <tr>
          <td
            class="Cell Cell-lit"
            id="0-0"
          />
          <td
            class="Cell Cell-lit"
            id="0-1"
          />
          <td
            class="Cell Cell-lit"
            id="0-2"
          />
          <td
            class="Cell Cell-lit"
            id="0-3"
          />
          <td
            class="Cell Cell-lit"
            id="0-4"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            id="1-0"
          />
          <td
            class="Cell Cell-lit"
            id="1-1"
          />
          <td
            class="Cell Cell-lit"
            id="1-2"
          />
          <td
            class="Cell Cell-lit"
            id="1-3"
          />
          <td
            class="Cell Cell-lit"
            id="1-4"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            id="2-0"
          />
          <td
            class="Cell Cell-lit"
            id="2-1"
          />
          <td
            class="Cell Cell-lit"
            id="2-2"
          />
          <td
            class="Cell Cell-lit"
            id="2-3"
          />
          <td
            class="Cell Cell-lit"
            id="2-4"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            id="3-0"
          />
          <td
            class="Cell Cell-lit"
            id="3-1"
          />
          <td
            class="Cell Cell-lit"
            id="3-2"
          />
          <td
            class="Cell Cell-lit"
            id="3-3"
          />
          <td
            class="Cell Cell-lit"
            id="3-4"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            id="4-0"
          />
          <td
            class="Cell Cell-lit"
            id="4-1"
          />
          <td
            class="Cell Cell-lit"
            id="4-2"
          />
          <td
            class="Cell Cell-lit"
            id="4-3"
          />
          <td
            class="Cell Cell-lit"
            id="4-4"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
